Tests/kaas/k8s-default-storage-class/k8s-default-storage-class-check.py

```diff
@@ -180,28 +184,139 @@ def check_default_persistentvolumeclaim_readwriteonce(k8s_api_instance, storage_
-    logger.debug(f"delete pvc:{pvc_name}")
-    api_response = k8s_api_instance.delete_namespaced_persistent_volume_claim(
-        pvc_name, namespace
-    )
 
+def check_csi_provider(k8s_core_api, allowed_csi_prov=ALLOWED_CSI_PROV):
+    pods = k8s_core_api.list_namespaced_pod(namespace="kube-system")
+    csi_list = []
+    for pod in pods.items:
+        if "csi" in pod.metadata.name:
+            if any(provider in pod.metadata.name for provider in allowed_csi_prov):
+                csi_list.append(pod.metadata.name)
+                logger.info(f"CSI-Provider: {pod.metadata.name}")
+            else:
+                raise SCSTestException(
+                    f"CSI-Provider: {pod.metadata.name} not recommended",
+                    return_code=33,
+                )
+    if not csi_list:
+      logger.info("CSI-Provider: No CSI Provider found.")
     return 0
 
 
-def main(argv):
+class TestEnvironment:
+    def __init__(self, kubeconfig):
+        self.namespace = NAMESPACE
+        self.pod_name = POD_NAME
+        self.pvc_name = PVC_NAME
+        self.k8s_core_api = None
+        self.return_code = 0
+        self.return_message = "return_message: FAILED"
+        self.kubeconfig = kubeconfig
+        self.cleanup = False
+
+    def prepare(self):
+        """
+        Sets up k8s client in preparation of the testing.
+        Checks whether the test pod and/or pvc already exists in the test namespace.
+        If so, it returns True indecating cleanup has to be executed first.
+        """
+        try:
+            logger.debug("setup_k8s_client(kubeconfig)")
+            self.k8s_core_api, self.k8s_storage_api = setup_k8s_client(self.kubeconfig)
+        except Exception as exception_message:
+            logger.error(f"L{inspect.currentframe().f_lineno} {exception_message}")
+            self.return_message = f"{exception_message}"
+            self.return_code = 1
+
+        logger.debug("Checking Environment for Leftovers")
+        try:
+            pod_list = self.k8s_core_api.list_namespaced_pod(namespace=self.namespace)
+            for pod in pod_list.items:
+                if pod.metadata.name == self.pod_name:
+                    logger.debug(
+                        f"POD '{self.pod_name}' exists in namespace '{self.namespace}'"
+                    )
+                    return True
+            pvc_list = self.k8s_core_api.list_namespaced_persistent_volume_claim(
+                namespace=self.namespace
+            )
+            for pvc in pvc_list.items:
+                if pvc.metadata.name == self.pvc_name:
+                    logger.debug(
+                        f"PVC '{self.pvc_name}' exists in namespace '{self.namespace}'"
+                    )
+                    return True
+            return False
+        except ApiException as e:
+            logger.error(f"Error preparing Environment: {e}")
+            return False
+
+    def clean(self):
```

Comment:
I think we can get rid of (big parts of) `prepare` as well as the variable `cleanup` by making this method ignore any error that says "resource not found". To me, this would be very much preferable. (It would in principle also reduce the risk of race conditions.)

Reply:
this is a good idea

Reply:
well, i can catch a "404 resource not found"  now and then skip the clean() execution, but if only one of both (pvc,pod) is missing the other one does not get deleted. so i am thinking, that i should ratehr catch that in the clean() function itself. what do you think?

Reply:
My proposal was to call `delete_namespace_pod` and `delete_namespaced_persistent_volume_claim` unconditionally, and to swallow any exception that indicates that deletion was unsuccessful because the resource didn't exist in the first place.

Reply:
i have done so now.. this leads to another problem: when the 404 is triggered, it doesn' t distinguish between the resources that are not found, so if e.g. the pvc is already built and only the pod is not found it will not delete the pvc.. so i will need to establish a flag